onnxruntime/core/graph/contrib_ops/collective_defs.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/graph/contrib_ops/contrib_defs.h"
+#include "core/graph/constants.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+using ONNX_NAMESPACE::AttributeProto;
+using ONNX_NAMESPACE::InferenceContext;
+using ONNX_NAMESPACE::OpSchema;
+using ONNX_NAMESPACE::OPTIONAL_VALUE;
+using ONNX_NAMESPACE::TypeProto;
+
+void RegisterCollectiveOps() {
+  ONNX_CONTRIB_OPERATOR_SCHEMA(AllReduce)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "input", "tensors to be reduced", "T", OpSchema::Variadic)
+      .Output(0, "output", "reduced tensors", "T", OpSchema::Variadic)
+      .TypeConstraint(
+          "T",
+          {"tensor(float16)", "tensor(float)", "tensor(double)"},
+          "Constrain to float, float16 and double tensors.")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        propagateShapeAndTypeFromFirstInput(ctx);
+      });
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(AllGather)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Attr("world_size",
+            "total world size that need to be gathered.",
+            AttributeProto::INT,
+            static_cast<int64_t>(1))
+      .Input(0, "input", "tensors to be sent", "T", OpSchema::Variadic)
+      .Output(0, "output", "gathered tensors", "T", OpSchema::Variadic)
+      .TypeConstraint(
+          "T",
+          {"tensor(float16)", "tensor(float)", "tensor(double)"},
+          "Constrain to float, float16 and double tensors.")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        auto world_size = getAttribute(ctx, "world_size", 1);
+        assert(world_size >= static_cast<int64_t>(1));
+        // propagate type for output
+        propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+        // propagate shape for output.
+        // output shape is [world_size * input_shape[0], ...]
+        auto output_type = ctx.getOutputType(0);
+        auto input_type = ctx.getInputType(0);
+        // all input/output should be tensor
+        assert(input_type->value_case() == output_type->value_case() && input_type->value_case() == TypeProto::kTensorType);
```

Comment:
nit: The assert seems redundant given that the type is propagated just a few lines above and given the type-constraints above.